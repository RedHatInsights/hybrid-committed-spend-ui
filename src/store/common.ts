import type { MapDispatchToProps, MapStateToProps } from 'react-redux';
import type { AnyAction } from 'redux';
import type { ThunkAction as ReduxThunkAction } from 'redux-thunk';

import type { RootState } from './rootReducer';

export const expirationMS = 30 * 60 * 1000; // 30 minutes

// eslint-disable-next-line no-shadow
export const enum FetchStatus {
  'none',
  'inProgress',
  'complete',
}

export function createMapStateToProps<OwnProps, StateProps>(
  mapStateToProps: MapStateToProps<StateProps, OwnProps, RootState>
) {
  return mapStateToProps;
}

export function createMapDispatchToProps<OwnProps, DispatchProps>(
  mapDispatchToProps: MapDispatchToProps<DispatchProps, OwnProps>
) {
  return mapDispatchToProps;
}

export type ThunkAction<A extends AnyAction = any, R = any> = ReduxThunkAction<R, RootState, void, A>;
