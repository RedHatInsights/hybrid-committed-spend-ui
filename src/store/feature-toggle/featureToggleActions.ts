import { createAction } from 'typesafe-actions';

export interface FeatureToggleActionMeta {
  isBillingStageToggleEnabled?: boolean;
  isDebugToggleEnabled?: boolean;
}

export const setFeatureToggle = createAction('feature/init_feature_toggle')<FeatureToggleActionMeta>();
export const resetState = createAction('feature/reset_state')();
