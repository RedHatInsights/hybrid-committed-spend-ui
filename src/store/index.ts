export { RootState } from './rootReducer';
export { configureStore } from './store';
