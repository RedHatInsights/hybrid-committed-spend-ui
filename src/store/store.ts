import notificationsMiddleware from '@redhat-cloud-services/frontend-components-notifications/notificationsMiddleware';
import axios from 'axios';
import type { DeepPartial } from 'redux';
import { applyMiddleware, compose, createStore } from 'redux';
import thunk from 'redux-thunk';

import type { RootState } from './rootReducer';
import { rootReducer } from './rootReducer';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__(options: any): typeof compose;
  }
}

const composeEnhancers =
  typeof window === 'object' && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ serialize: true })
    : compose;

export const middlewares = [thunk, notificationsMiddleware()];

export function configureStore(initialState: DeepPartial<RootState>) {
  const enhancer = composeEnhancers(applyMiddleware(...middlewares));
  const store = createStore(rootReducer, initialState as any, enhancer);

  axios.interceptors.response.use(null, error => {
    return Promise.reject(error);
  });

  return store;
}
