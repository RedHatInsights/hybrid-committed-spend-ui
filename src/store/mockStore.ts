import type { DeepPartial, Reducer, ReducersMapObject, Store } from 'redux';
import { applyMiddleware, combineReducers, createStore } from 'redux';

import type { RootState } from './rootReducer';
import { middlewares } from './store';

type MockReducer<S> = ReducersMapObject<S, any> | Reducer<S>;

export function createMockStoreCreator<S>(reducer: MockReducer<S>) {
  const rootReducer = typeof reducer === 'object' ? combineReducers(reducer) : reducer;

  return (initialState?: DeepPartial<S>): Store<RootState, any> => {
    return createStore<RootState, any, any, any>(rootReducer as any, initialState, applyMiddleware(...middlewares));
  };
}
