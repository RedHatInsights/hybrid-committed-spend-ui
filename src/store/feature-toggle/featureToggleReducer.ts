import type { ActionType } from 'typesafe-actions';
import { getType } from 'typesafe-actions';

import type { resetState } from './featureToggleActions';
import { setFeatureToggle } from './featureToggleActions';

export type FeatureToggleAction = ActionType<typeof setFeatureToggle | typeof resetState>;

export type FeatureToggleState = Readonly<{
  hasFeatureToggle: boolean;
  isBillingStageToggleEnabled: boolean;
  isDebugToggleEnabled: boolean;
}>;

export const defaultState: FeatureToggleState = {
  hasFeatureToggle: false,
  isBillingStageToggleEnabled: false,
  isDebugToggleEnabled: false,
};

export const stateKey = 'featureToggle';

export function featureToggleReducer(state = defaultState, action: FeatureToggleAction): FeatureToggleState {
  switch (action.type) {
    case getType(setFeatureToggle):
      return {
        ...state,
        hasFeatureToggle: true,
        isBillingStageToggleEnabled: action.payload.isBillingStageToggleEnabled,
        isDebugToggleEnabled: action.payload.isDebugToggleEnabled,
      };

    default:
      return state;
  }
}
