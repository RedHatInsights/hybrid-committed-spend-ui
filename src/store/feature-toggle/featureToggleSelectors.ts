import type { RootState } from 'store/rootReducer';

import { stateKey } from './featureToggleReducer';

export const selectFeatureToggleState = (state: RootState) => state[stateKey];

export const selectHasFeatureToggle = (state: RootState) => selectFeatureToggleState(state).hasFeatureToggle;

export const selectIsBillingStageToggleEnabled = (state: RootState) =>
  selectFeatureToggleState(state).isBillingStageToggleEnabled;
export const selectIsDebugToggleEnabled = (state: RootState) => selectFeatureToggleState(state).isDebugToggleEnabled;
