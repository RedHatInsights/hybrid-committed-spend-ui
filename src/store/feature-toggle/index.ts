import * as featureToggleActions from './featureToggleActions';
import type { FeatureToggleAction, FeatureToggleState } from './featureToggleReducer';
import { featureToggleReducer, stateKey as featureToggleStateKey } from './featureToggleReducer';
import * as featureToggleSelectors from './featureToggleSelectors';

export { featureToggleActions, featureToggleReducer, featureToggleSelectors, featureToggleStateKey };
export type { FeatureToggleAction, FeatureToggleState };
