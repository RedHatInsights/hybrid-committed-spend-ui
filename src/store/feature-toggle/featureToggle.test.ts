import { createMockStoreCreator } from 'store/mockStore';

import { featureToggleSelectors } from '.';
import * as actions from './featureToggleActions';
import { featureToggleReducer, stateKey } from './featureToggleReducer';
import * as selectors from './featureToggleSelectors';

const createUIStore = createMockStoreCreator({
  [stateKey]: featureToggleReducer,
});

test('default state', async () => {
  const store = createUIStore();
  expect(selectors.selectFeatureToggleState(store.getState())).toMatchSnapshot();
});

test('Billing stage feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureToggle({ isBillingStageToggleEnabled: true }));
  expect(featureToggleSelectors.selectIsBillingStageToggleEnabled(store.getState())).toBe(true);
});
