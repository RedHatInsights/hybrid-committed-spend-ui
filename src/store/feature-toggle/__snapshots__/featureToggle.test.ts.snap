// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default state 1`] = `
{
  "hasFeatureToggle": false,
  "isBillingStageToggleEnabled": false,
  "isDebugToggleEnabled": false,
}
`;
